core/sdk-core/src/main/java/software/amazon/awssdk/core/async/AsyncRequestBody.java

```diff
@@ -143,29 +144,96 @@ static AsyncRequestBody fromString(String string) {
      */
     static AsyncRequestBody fromBytes(byte[] bytes) {
-        return new ByteArrayAsyncRequestBody(bytes, Mimetype.MIMETYPE_OCTET_STREAM);
+        byte[] clonedBytes = bytes.clone();
+        return ByteBuffersAsyncRequestBody.from(clonedBytes);
     }
 
     /**
-     * Creates a {@link AsyncRequestBody} from a {@link ByteBuffer}. Buffer contents are copied so any modifications
-     * made to the original {@link ByteBuffer} are not reflected in the {@link AsyncRequestBody}.
+     * Creates an {@link AsyncRequestBody} from a byte array <b>without</b> copying the contents of the byte array. This
+     * introduces concurrency risks, allowing: (1) the caller to modify the byte array stored in this {@code AsyncRequestBody}
+     * implementation AND (2) any users of {@link #fromBytesUnsafe(byte[])} to modify the byte array passed into this
+     * {@code AsyncRequestBody} implementation.
+     *
+     * <p>As the method name implies, this is unsafe. Use {@link #fromBytes(byte[])} unless you're sure you know the risks.
+     *
+     * @param bytes The bytes to send to the service.
+     * @return AsyncRequestBody instance.
+     */
+    static AsyncRequestBody fromBytesUnsafe(byte[] bytes) {
+        return ByteBuffersAsyncRequestBody.from(bytes);
+    }
+
+    /**
+     * Creates an {@link AsyncRequestBody} from a {@link ByteBuffer}. This will copy the contents of the {@link ByteBuffer} to
+     * prevent modifications to the provided {@link ByteBuffer} from being reflected in the {@link AsyncRequestBody}.
+     * <p> The position is set to 0 in the copied {@link ByteBuffer} and the mark if defined is discarded.
      *
      * @param byteBuffer ByteBuffer to send to the service.
      * @return AsyncRequestBody instance.
      */
     static AsyncRequestBody fromByteBuffer(ByteBuffer byteBuffer) {
-        return fromBytes(BinaryUtils.copyAllBytesFrom(byteBuffer));
+        ByteBuffer immutableCopy = BinaryUtils.immutableCopyOf(byteBuffer);
+        immutableCopy.rewind();
+        return ByteBuffersAsyncRequestBody.of(null, immutableCopy);
+    }
+
+    /**
+     * Creates an {@link AsyncRequestBody} from a {@link ByteBuffer} <b>without</b> copying the contents of the
+     * {@link ByteBuffer}. This introduces concurrency risks, allowing: (1) the caller to modify the {@link ByteBuffer} stored in
+     * this {@code AsyncRequestBody} implementation AND (2) any users of {@link #fromByteBufferUnsafe(ByteBuffer)} to modify the
+     * {@link ByteBuffer} passed into this {@code AsyncRequestBody} implementation.
+     *
+     * <p>As the method name implies, this is unsafe. Use {@link #fromByteBuffer(ByteBuffer)}} unless you're sure you know the
+     * risks.
+     *
+     * @param byteBuffer ByteBuffer to send to the service.
+     * @return AsyncRequestBody instance.
+     */
+    static AsyncRequestBody fromByteBufferUnsafe(ByteBuffer byteBuffer) {
+        return ByteBuffersAsyncRequestBody.of(null, byteBuffer);
+    }
```

Comment:
Sorry for the additional comments, since it felt like we were done, but I've just realized that semantics of these two seem to be different for the position. It seems like fromByteBuffer starts from position 0 (for some reason) and fromByteBufferUnsafe does not (which seems more correct).

Should we do the (probably less correct) thing for `fromByteBufferUnsafe` with `byteBuffer.asReadOnlyBuffer().rewind()`? (same with `fromByteBuffersUnsafe`)

Later on, we should probably deprecate `fromByteBuffer` for a new (more correct) `fromByteBufferPosition` and `fromByteBufferPositionUnsafe` that actually honors the current position of the buffer.

Reply:
Yes I noticed this too, looking at [`RequestBody`](https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/sync/RequestBody.java#L174-L198) I think adding the implementation for `fromRemainingByteBuffer`, `fromRemainingByteBufferUnsafe` and `fromRemainingByteBufferUnsafe` makes sense.

Reply:
I think this method ensuring it passes along a read only buffer makes sense too to mitigate
>(2) any users of {@link #fromByteBufferUnsafe(ByteBuffer)} to modify the
     * {@link ByteBuffer} passed into this {@code AsyncRequestBody} implementation.

Reply:
I've made the changes and kept them in a [separate commit](https://github.com/aws/aws-sdk-java-v2/pull/3925/commits/1fd1420bfd97d796d2c90f846e61ff7c6e87be0a) to make it easier to review, if you're happy I'll merge these commits.

Reply:
LGTM, let's merge!

Reply:
commits merged 🙂